### PR TITLE
Fileservice issue82

### PR DESCRIPTION
**Pull request by serchch**
FileService Basic Structure, Contracts, Services
#82

### CODE REVIEW

File: src/FileServiceManager/FileService/Startup.cs [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                options.IgnoreObsoleteActions();
+                options.DescribeAllParametersInCamelCase();
+                options.SwaggerDoc("v1", new OpenApiInfo
+                {
+                    Title = "Travely.FileService API",
+                    Version = "v1",
+                    TermsOfService = new System.Uri("https://github.com/titancamp/travely"),
+                    Contact = new OpenApiContact { Name = "https://www.servicetitan.com" }
+                });
+                //options.AddSecurityDefinition("JWT", new OpenApiSecurityScheme
```
Comment (TDavtyan-ST) [inline src/FileServiceManager/FileService/Startup.cs:40]: do not leave unused code in the committed code

File: src/FileServiceManager/Models/FileMetadata.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+namespace FileService.Models
+{
+    public class FileMetadata
+    {
+        public Guid Id { get; set; }
+        public string Name { get; set; }
+        public string Extension { get; set; }
+        public string UploadedBy { get; set; }
+        public DateTime? CreatedOn { get; set; }
```
Comment (TDavtyan-ST) [inline src/FileServiceManager/Models/FileMetadata.cs:11]: why CreatedOn property is nullable ?

File: src/FileServiceManager/Models/FileMetadata.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+        public string FilePath { get; set; }
+        public byte[] Content { get; set; }
```
Comment (TDavtyan-ST) [inline src/FileServiceManager/Models/FileMetadata.cs:13]: Metadata is about everything , but not the content itself

File: src/FileServiceManager/FileService/DAL/Storages/FileSystemStorage.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+using Microsoft.Extensions.Configuration;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace FileService.DAL
+{
+    public class FileSystemStorage : IStorage
+    {
+        private readonly IConfiguration _configuration;
```
Comment (TDavtyan-ST) [inline src/FileServiceManager/FileService/DAL/Storages/FileSystemStorage.cs:12]: do we need this configuration here ?

File: src/FileServiceManager/FileService/DAL/Interfaces/IStorage.cs [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+using System.IO;
+using System.Threading.Tasks;
+
+namespace FileService.DAL
+{
+    public interface IStorage
+    {
+        Task<Guid> UploadFileAsync(IFormFile file, int companyId);
+
+        Task<IEnumerable<FileMetadata>> GetAllFilesAsync(int companyId);
```
Comment (TDavtyan-ST) [inline src/FileServiceManager/FileService/DAL/Interfaces/IStorage.cs:14]: getting all files does not mean downloading them all, please consider excluding the content of file from the FileMetada class

File: src/FileServiceManager/FileService/Controllers/FileController.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+        public FileController(IStorage storage)
+        {
+            _storage = storage;
+        }
+
+        [HttpGet("Download/{fileId}")]
+        public async Task<FileMetadata> DownLoadFileAsync(Guid fileId, int companyId)
+        {
+            //identity check
```
Comment (TDavtyan-ST) [inline src/FileServiceManager/FileService/Controllers/FileController.cs:28]: think of this identity check as an action filter, to not have the same logic written in every method

File: src/FileServiceManager/FileService/DAL/Interfaces/IStorage.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+        Task<FileMetadata> DownLoadFileAsync(Guid fileId, int companyId);
```
Comment (TDavtyan-ST) [inline src/FileServiceManager/FileService/DAL/Interfaces/IStorage.cs:16]: download should open browsers download experience, please consider changing this method signature and the implementation in future appropriately

File: src/FileServiceManager/FileService/DAL/Storages/FileSystemStorage.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+        public Task<bool> RemoveFileAsync(Guid fileId, int companyId)
+        {
+            throw new NotImplementedException();
+        }
+
+        public async Task<Guid> UploadFileAsync(IFormFile file, int companyId)
+        {
+            throw new NotImplementedException();
+            //var basePath = Path.Combine(Directory.GetCurrentDirectory() + "\\Files\\");
```
Comment (TDavtyan-ST) [inline src/FileServiceManager/FileService/DAL/Storages/FileSystemStorage.cs:36]: do not leave unused code in committed code

File: src/FileServiceManager/FileService/Helpers/RESTException.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Runtime.Serialization;
+using System.Security.Permissions;
+using System.Net;
+
+namespace FileService.Helpers
+{
+    public class RESTException : Exception
```
Comment (TDavtyan-ST) [inline src/FileServiceManager/FileService/Helpers/RESTException.cs:8]: didn't saw the usage of this class, is it for future use ?

**Review by TDavtyan-ST — CHANGES_REQUESTED** [review]
Go over the comments and let us discuss them

File: src/FileServiceManager/FileService/Startup.cs (hunk 1 above)
Comment (serchch) [reply src/FileServiceManager/FileService/Startup.cs:40]: Fixed

File: src/FileServiceManager/FileService/Startup.cs [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+            switch (Configuration["storage:type"])
+            {
+                case "FileSystem":
+                    services.AddScoped<IStorage, FileSystemStorage>();
+                    break;
+                case "AzureBlob":
+                    //services.AddScoped<IStorage, AzureBlobStorage>();
+                    break;
+                default:
+                    services.AddScoped<IStorage, FileSystemStorage>();
```
[the remark this answers is not in the document]
Comment (serchch) [reply src/FileServiceManager/FileService/Startup.cs:59]: This is the case when this part of config is empty. Maybe we can add other Storage type.

File: src/FileServiceManager/Models/FileMetadata.cs (hunk 2 above)
Comment (serchch) [reply src/FileServiceManager/Models/FileMetadata.cs:11]: It is my mistake. Fixed

File: src/FileServiceManager/Models/FileMetadata.cs (hunk 3 above)
Comment (serchch) [reply src/FileServiceManager/Models/FileMetadata.cs:13]: There was misunderstanding during our previous conversations. After last call, it becomes more clear. I remove content.

File: src/FileServiceManager/FileService/DAL/Storages/FileSystemStorage.cs (hunk 4 above)
Comment (serchch) [reply src/FileServiceManager/FileService/DAL/Storages/FileSystemStorage.cs:12]: Removed

File: src/FileServiceManager/FileService/Controllers/FileController.cs (hunk 6 above)
Comment (serchch) [reply src/FileServiceManager/FileService/Controllers/FileController.cs:28]: I remove the comments. The implementation will follow

File: src/FileServiceManager/FileService/DAL/Interfaces/IStorage.cs (hunk 5 above)
Comment (serchch) [reply src/FileServiceManager/FileService/DAL/Interfaces/IStorage.cs:14]: Fixed

File: src/FileServiceManager/FileService/DAL/Storages/FileSystemStorage.cs (hunk 8 above)
Comment (serchch) [reply src/FileServiceManager/FileService/DAL/Storages/FileSystemStorage.cs:36]: Fixed

File: src/FileServiceManager/FileService/Helpers/RESTException.cs (hunk 9 above)
Comment (serchch) [reply src/FileServiceManager/FileService/Helpers/RESTException.cs:8]: I supposed to use it in future as one type for all type exceptions. We can map all exceptions to this exception type and let the client to "parse" our exception. Anyway, currently I remove it till the last decision.

File: src/FileServiceManager/FileService/DAL/Interfaces/IStorage.cs (hunk 7 above)
Comment (serchch) [reply src/FileServiceManager/FileService/DAL/Interfaces/IStorage.cs:16]: I've changed to FileResult as per our conversation

**Review by TDavtyan-ST — APPROVED** [review]

**Pull request merged**